File: backend/app/models/factor.py
"""Factor result/config models for SW sector momentum."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Index,
    Numeric,
    String,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base, mapped_col_created_at, mapped_col_updated_at


class FactorConfig(Base):
    __tablename__ = "factor_config"
    __table_args__ = (
        UniqueConstraint("name", "owner", name="uq_factor_config_name_owner"),
        Index("ix_factor_config_owner_updated", "owner", "updated_at"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    params: Mapped[dict] = mapped_column(JSONB, nullable=False)
    owner: Mapped[str] = mapped_column(String(64), nullable=False)
    created_at: Mapped[datetime] = mapped_col_created_at()
    updated_at: Mapped[datetime] = mapped_col_updated_at()
    updated_by: Mapped[str] = mapped_column(String(64), nullable=False)


class FactorResult(Base):
    __tablename__ = "factor_result"
    __table_args__ = (
        Index("ix_factor_result_created", "created_at"),
        Index("ix_factor_result_basedate_created", "basedate", "created_at"),
        Index("ix_factor_result_class_basedate", "classification", "basedate"),
        Index("ix_factor_result_stale", "stale"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    params: Mapped[dict] = mapped_column(JSONB, nullable=False)
    basedate: Mapped[date] = mapped_column(Date, nullable=False)
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    classification: Mapped[str] = mapped_column(String(16), nullable=False)
    industry_snapshot_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    stale: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    stale_reason: Mapped[str | None] = mapped_column(String(64), nullable=True)
    stale_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    created_by: Mapped[str] = mapped_column(String(64), nullable=False)


class FactorResultRow(Base):
    __tablename__ = "factor_result_row"
    __table_args__ = (
        UniqueConstraint("result_id", "level", "sector_code", name="uq_factor_row_result_level_sector"),
        Index("ix_factor_row_result_level_score", "result_id", "level", "momentum_score"),
        Index("ix_factor_row_result_parent", "result_id", "parent_sector_code"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    result_id: Mapped[int] = mapped_column(
        ForeignKey("factor_result.id", ondelete="CASCADE"), nullable=False
    )
    level: Mapped[str] = mapped_column(String(4), nullable=False)
    sector_code: Mapped[str] = mapped_column(String(32), nullable=False)
    sector_name: Mapped[str] = mapped_column(String(128), nullable=False)
    parent_sector_code: Mapped[str | None] = mapped_column(String(32), nullable=True)
    sector_stock_count: Mapped[int] = mapped_column(nullable=False)
    sector_top_stock_count: Mapped[int] = mapped_column(nullable=False)
    top_density: Mapped[Decimal] = mapped_column(Numeric(10, 6), nullable=False)
    median_return: Mapped[Decimal | None] = mapped_column(Numeric(12, 6), nullable=True)
    momentum_score: Mapped[Decimal] = mapped_column(Numeric(14, 6), nullable=False)
    small_sample_flag: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)


class FactorResultStock(Base):
    __tablename__ = "factor_result_stock"
    __table_args__ = (
        UniqueConstraint("result_id", "ts_code", name="uq_factor_stock_result_ts_code"),
        Index("ix_factor_stock_result_l1_return", "result_id", "l1_code", "stock_return"),
        Index("ix_factor_stock_result_l2_return", "result_id", "l2_code", "stock_return"),
        Index("ix_factor_stock_result_l3_return", "result_id", "l3_code", "stock_return"),
        Index("ix_factor_stock_result_top", "result_id", "is_top"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    result_id: Mapped[int] = mapped_column(
        ForeignKey("factor_result.id", ondelete="CASCADE"), nullable=False
    )
    ts_code: Mapped[str] = mapped_column(String(16), nullable=False)
    stock_name: Mapped[str] = mapped_column(String(64), nullable=False)
    l1_code: Mapped[str | None] = mapped_column(String(32), nullable=True)
    l1_name: Mapped[str | None] = mapped_column(String(128), nullable=True)
    l2_code: Mapped[str | None] = mapped_column(String(32), nullable=True)
    l2_name: Mapped[str | None] = mapped_column(String(128), nullable=True)
    l3_code: Mapped[str | None] = mapped_column(String(32), nullable=True)
    l3_name: Mapped[str | None] = mapped_column(String(128), nullable=True)
    stock_return: Mapped[Decimal | None] = mapped_column(Numeric(14, 8), nullable=True)
    is_top: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    missing_reason: Mapped[str | None] = mapped_column(String(64), nullable=True)

File: frontend/src/api/factor.ts
import { apiDelete, apiGet, apiPatch, apiPost } from "@/api/http";

export type FactorLevel = "L1" | "L2" | "L3";

export interface FactorParams {
  basedate: string;
  window: number;
  top_ratio: number;
  classification: "SW";
  level: FactorLevel;
  return_method: "simple" | "log";
  score_method: "median_return_score" | "top_count_score";
}

export interface FactorResultHead {
  id: number;
  params: FactorParams;
  basedate: string;
  start_date: string;
  classification: string;
  industry_snapshot_at: string | null;
  stale: boolean;
  stale_reason: string | null;
  created_at: string;
  created_by: string;
}

export interface FactorRow {
  id: number;
  result_id: number;
  level: FactorLevel;
  sector_code: string;
  sector_name: string;
  parent_sector_code: string | null;
  sector_stock_count: number;
  sector_top_stock_count: number;
  top_density: number;
  median_return: number | null;
  momentum_score: number;
  small_sample_flag: boolean;
}

export interface FactorStockRow {
  ts_code: string;
  stock_name: string;
  stock_return: number | null;
  is_top: boolean;
  missing_reason: string | null;
}

export interface FactorConfig {
  id: number;
  name: string;
  params: FactorParams;
  updated_at: string;
}

export function calculateFactor(params: FactorParams): Promise<{ result: FactorResultHead; level: FactorLevel; rows: FactorRow[] }> {
  return apiPost("/api/factor/results", params);
}

export function getFactorResult(id: number, level?: FactorLevel): Promise<{ result: FactorResultHead; level: FactorLevel; rows: FactorRow[] }> {
  return apiGet(`/api/factor/results/${id}`, level ? { level } : undefined);
}

export function getFactorResults(): Promise<{ items: FactorResultHead[] }> {
  return apiGet("/api/factor/results");
}

export function getFactorChildren(id: number, parent_level: FactorLevel, parent_sector_code: string): Promise<{ level: FactorLevel; rows: FactorRow[] }> {
  return apiGet(`/api/factor/results/${id}/children`, { parent_level, parent_sector_code });
}

export function getFactorSectorStocks(id: number, level: FactorLevel, sectorCode: string): Promise<{ stocks: FactorStockRow[] }> {
  return apiGet(`/api/factor/results/${id}/sectors/${level}/${sectorCode}/stocks`);
}

export function recalculateFactor(id: number): Promise<{ result: FactorResultHead; level: FactorLevel; rows: FactorRow[] }> {
  return apiPost(`/api/factor/results/${id}/recalculate`);
}

export function getFactorConfigs(): Promise<{ items: FactorConfig[] }> {
  return apiGet("/api/factor/configs");
}

export function createFactorConfig(body: { name: string; params: FactorParams }): Promise<FactorConfig> {
  return apiPost("/api/factor/configs", body);
}

export function updateFactorConfig(id: number, body: Partial<{ name: string; params: FactorParams }>): Promise<FactorConfig> {
  return apiPatch(`/api/factor/configs/${id}`, body);
}

export function copyFactorConfig(id: number): Promise<FactorConfig> {
  return apiPost(`/api/factor/configs/${id}/copy`);
}

export function deleteFactorConfig(id: number): Promise<{ deleted: boolean }> {
  return apiDelete(`/api/factor/configs/${id}`);
}
